traversal/planner/GraphPlanner.java

```diff
@@ -65,9 +68,9 @@ public class GraphPlanner implements Planner {
     private final Map<Identifier, PlannerVertex<?>> vertices;
     private final Set<PlannerEdge<?, ?>> edges;
     private final AtomicBoolean isOptimising;
-    private final ReadWriteLock firstOptimisingLock;
 
     protected volatile GraphProcedure procedure;
+    private volatile CompletableFuture<Void> backgroundOptimisation;
```

Comment:
background planning task that may or may not be completed at any given time but can always be waited on with a time limit